src/Microsoft.CSharp/tests/ArrayHandling.cs

```diff
@@ -0,0 +1,55 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using Xunit;
+
+namespace Microsoft.CSharp.RuntimeBinder.Tests
+{
+    public class ArrayHandling
+    {
+        [Fact]
+        public void SingleRankNonSZArray()
+        {
+            dynamic d = Array.CreateInstance(typeof(int), new[] { 8 }, new[] { -2 });
+            d.SetValue(32, 3);
+            d.SetValue(28, -1);
+            Assert.Equal(32, d.GetValue(3));
+            Assert.Equal(28, d.GetValue(-1));
+        }
+
+        [Fact]
+        public void SingleRankNonSZArrayIndexed()
+        {
+            dynamic d = Array.CreateInstance(typeof(int), new[] { 8 }, new[] { -2 });
+            d[3] = 32;
+            d[-1] = 28;
+            Assert.Equal(32, d[3]);
+            Assert.Equal(28, d[-1]);
```

Comment:
Also, one can currently do the equivalent with non-zero-based arrays of higher ranks, so prohibiting it here would have to either be a breaking change to that or else inconsistent with it.